### PR TITLE
Setup vsphere cluster reconciler for new api

### PR DESCRIPTION
**Pull request by taneyland**
*Issue #, if available:*

*Description of changes:*
The vsphere cluster reconciler already exists & is registered. This PR removes some unnecessary code & sets up the vsphere reconciler for the forthcoming changes for the new API

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 1]
```diff
@@ -113,181 +96,57 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
-		}
+func (r *Reconciler) validateMachineConfigs(ctx context.Context, cluster *anywherev1.Cluster, clusterSpec *c.Spec, datacenterConfig *anywherev1.VSphereDatacenterConfig) error {
+	machineConfigMap := map[string]*anywherev1.VSphereMachineConfig{}
 
-		if err := serverside.ReconcileYaml(ctx, r.client, workersSpec); err != nil {
-			return controller.Result{}, err
+	for _, ref := range cluster.MachineConfigRefs() {
+		machineConfig := &anywherev1.VSphereMachineConfig{}
+		machineConfigName := types.NamespacedName{Namespace: cluster.Namespace, Name: ref.Name}
+		if err := r.client.Get(ctx, machineConfigName, machineConfig); err != nil {
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:140]: The machine configs are already in the cluster spec, we can just get them from there

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 2]
```diff
@@ -113,181 +96,57 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
 	log.V(4).Info("Fetching cluster spec")
 	specWithBundles, err := c.BuildSpec(ctx, clientutil.NewKubeClient(r.client), cluster)
 	if err != nil {
 		return controller.Result{}, err
 	}
 
-	vsphereClusterSpec := vsphere.NewSpec(specWithBundles, machineConfigMap, dataCenterConfig)
-
-	if err := r.validator.ValidateClusterMachineConfigs(ctx, vsphereClusterSpec); err != nil {
+	if err := r.validateMachineConfigs(ctx, cluster, specWithBundles, dataCenterConfig); err != nil {
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:105]: Could we make the validation another phase in the runner?

Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:105]: the datacenter is already in the cluster spec

Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:105]: Yes, with the other changes that you suggested, we can make this another phase

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 3]
```diff
@@ -113,181 +96,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
-		workersSpec, err := templateBuilder.GenerateCAPISpecWorkers(specWithBundles, workloadTemplateNames, kubeadmconfigTemplateNames)
-		if err != nil {
-			return controller.Result{}, err
-		}
-
-		if err := serverside.ReconcileYaml(ctx, r.client, workersSpec); err != nil {
-			return controller.Result{}, err
-		}
+func (r *Reconciler) ValidateMachineConfigs(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	_ = log.WithValues("phase", "validateMachineConfigs")
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:132]: if we are not using the log, no need to create one

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 4]
```diff
@@ -113,181 +96,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
-		if err != nil {
-			return controller.Result{}, err
-		}
-		if err := serverside.ReconcileYaml(ctx, r.client, controlPlaneSpec); err != nil {
-			return controller.Result{Result: &ctrl.Result{
-				RequeueAfter: defaultRequeueTime,
-			}}, err
-		}
-		conditions.MarkTrue(cluster, controlSpecPlaneAppliedCondition)
+	if err := r.validator.ValidateClusterMachineConfigs(ctx, vsphereClusterSpec); err != nil {
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:138]: should we bubble up this error to the cluster status?

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 5]
```diff
@@ -113,181 +96,47 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
-		machineConfig := &anywherev1.VSphereMachineConfig{}
-		machineConfigName := types.NamespacedName{Namespace: cluster.Namespace, Name: ref.Name}
-		if err := r.client.Get(ctx, machineConfigName, machineConfig); err != nil {
-			return controller.Result{}, err
-		}
-		machineConfigMap[ref.Name] = machineConfig
-	}
-
 	log.V(4).Info("Fetching cluster spec")
 	specWithBundles, err := c.BuildSpec(ctx, clientutil.NewKubeClient(r.client), cluster)
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:100]: can we rename this to just `spec`?

**Review by g-gaston — COMMENTED** [review]
A couple more things:
* Can we make the datacenter validation part a phase as well?
* I believe setupenvars is only needed for the validator, hence we can move that to the validate machine configs phase.

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 6]
```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
...
-	}
+func (r *Reconciler) ValidateDatacenterConfig(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "validateDatacenterConfig")
+	dataCenterConfig := clusterSpec.VSphereDatacenter
 
-	capiCluster, result, errCAPICLuster := r.getCAPICluster(ctx, cluster, log)
-	if errCAPICLuster != nil {
-		return result, errCAPICLuster
+	// Set up envs for executing Govc cmd and default values for datacenter config
+	if err := SetupEnvVars(ctx, dataCenterConfig, r.client); err != nil {
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:94]: this is not needed in this method, only in `ValidateMachineConfigs`

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 7]
```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
...
-		return result, err
-	}
+func (r *Reconciler) ValidateDatacenterConfig(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "validateDatacenterConfig")
+	dataCenterConfig := clusterSpec.VSphereDatacenter
 
-	capiCluster, result, errCAPICLuster := r.getCAPICluster(ctx, cluster, log)
-	if errCAPICLuster != nil {
-		return result, errCAPICLuster
+	// Set up envs for executing Govc cmd and default values for datacenter config
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:93]:
```suggestion
	// Set up envs for executing Govc cmd
```

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 8]
```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
...
-	if !conditions.IsTrue(capiCluster, controlPlaneReadyCondition) {
-		log.Info("waiting for control plane to be ready", "cluster", capiCluster.Name, "kind", capiCluster.Kind)
-		return controller.Result{Result: &ctrl.Result{
-			RequeueAfter: defaultRequeueTime,
-		}}, err
-	}
 
-	if result, err := r.reconcileCNI(ctx, cluster, capiCluster, specWithBundles, log); err != nil {
-		return result, err
+		log.Error(nil, failureMessage)
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:104]: maybe include the VSphereDatacenterConfig as a value in the log

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 9]
```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
...
-		ciliumSpec, err := cilium.GenerateManifest(ctx, specWithBundles, []string{constants.CapvSystemNamespace})
-		if err != nil {
-			return controller.Result{}, err
-		}
-		if err := serverside.ReconcileYaml(ctx, remoteClient, ciliumSpec); err != nil {
-			return controller.Result{}, err
-		}
-		conditions.MarkTrue(cluster, cniSpecAppliedCondition)
+	if err := r.validator.ValidateClusterMachineConfigs(ctx, vsphereClusterSpec); err != nil {
+		failureMessage := fmt.Sprintf("VSphereMachineConfig is invalid: %v", err)
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:119]:
```suggestion
		failureMessage := fmt.Sprintf("VSphereMachineConfig is invalid: %s", err)
```

File: pkg/providers/vsphere/reconciler/reconciler.go [hunk 10]
```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
...
+		log.Error(nil, failureMessage)
+		clusterSpec.Cluster.Status.FailureMessage = &failureMessage
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler.go:121]: are we testing all these stuff?

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 8 above)
Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:99]: I would prefer to keep the messaging for this as it currently is

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 9 above)
Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:120]: Should we not use `%v` for errors?

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 10 above)
Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:122]: yes, it's tested in [this file](https://github.com/aws/eks-anywhere/blob/main/controllers/cluster_controller_test.go) & I'm going to add more tests in my next PR that will include a `reconciler_test.go` file for vsphere reconciler

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 8 above)
Comment (g-gaston) [reply pkg/providers/vsphere/reconciler/reconciler.go:99]: Ok. Why?
If we don't add the datacenter, the message doesn't tell us which datacenter is invalid

```go
log.Error(errors.New(failureMessage), "Invalid VSphereDatacenterConfig, "datacenterConfig", klog.KObj(dataCenterConfig))
```

Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:99]: Oh we're adding the datacenterConfig name to the failure message a few lines above: https://github.com/aws/eks-anywhere/blob/72b11b3e59aa801154dbc554e063b8fdd7de761c/pkg/providers/vsphere/reconciler/reconciler.go#L94

But now i see your idea of just adding it as a log value. That works for me too!

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 9 above)
Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:120]: yes, makes sense!

File: pkg/providers/vsphere/reconciler/reconciler.go (hunk 10 above)
Comment (taneyland) [reply pkg/providers/vsphere/reconciler/reconciler.go:122]: will be sure to have tests with assertions to cover these cases!

**Review by g-gaston — APPROVED** [review]

**Review by g-gaston — APPROVED** [review]

**Pull request merged**